src/main/scala/Syntax.scala

```diff
@@ -112,12 +112,15 @@ object Syntax {
       case _:EVar | _:EArrAccess => true
       case _ => false
     }
+    var typ: Option[Type] = None
```

Comment:
What is this used for?

Reply:
I need to know the type of each index while bound checking. The type checking for this happens in `TypeChecker.consumeBanks` and so I save the type of each index expression.

Reply:
As usual, this opens up a bigger discussion. The general idea here is *type elaboration*, which is something most compilers need to do—i.e., associate a type with every AST node (or at least, with every expression and every declaration). It's necessary because most downstream passes need to know the types of everything. Code generation typically needs to know about types, and indeed that's why identifiers already have a type field (as Rachit mentioned some time ago: https://github.com/cucapra/seashell/pull/46/files#r254274169). And clearly bounds checking needs to know the types also.

So in this or some subsequent PR, we should address that problem head-on. That is, we should invent some data structure for holding the elaborated types, and both codegen and bounds checking should use that. It should replace the one-off type field for idents.

That raises the question about the right way to store the elaborated types. [This blog post by Edward Yang](http://blog.ezyang.com/2013/05/the-ast-typing-problem/) gives a succinct overview of the typical options, which are all imperfect in their own way. The `typ: Option[Type]` strategy is called "Nullable field" in the blog post.

Reply:
I think this would be better in its own PR. Another related topic is a nice way to write new passes over the tree. For example, for the bounds checking, I had to write a lot of uninteresting code as boilerplate:
```
  private def checkC(c: Command) : Unit = c match {
    case CPar(c1, c2) => checkC(c1) ; checkC(c2)
    case CSeq(c1, c2) => checkC(c1) ; checkC(c2)
    case CLet(_, _, exp) => checkE(exp)
    case CView(_, _) => ()
    case CIf(cond, tbranch, fbranch) => checkE(cond) ; checkC(tbranch) ; checkC(fbranch)
    case CFor(_, par, combine) => checkC(par) ; checkC(combine)
    case CWhile(cond, body) => checkE(cond) ; checkC(body)
    case CUpdate(lhs, rhs) => checkE(lhs) ; checkE(rhs)
    case CReduce(_, l, r) => checkE(l) ; checkE(r)
    case CExpr(e) => checkE(e)
    case CEmpty => ()
  }
```
It would be nice if I could simply target the branches where I want to do something.
 This depends a little bit on how many passes we anticipate writing.

Reply:
Yeah, that's definitely a common issue when writing compiler. In the OCaml codebase, I implemented [a visitor class](https://github.com/cucapra/seashell/blob/3d25f46e67f2512aba707bdf9daa9dc5d7c5f3aa/src/ast_visitor.ml) for the AST. That allowed us to write some [clean passes](https://github.com/cucapra/seashell/blob/3d25f46e67f2512aba707bdf9daa9dc5d7c5f3aa/src/flatten_seq.ml).

The code under RewriteViews.scala has beginning of a monadic visitor. It would be nice to extract the logic into something reusable.